src/metabase/lib/util.cljc

```diff
@@ -0,0 +1,100 @@
+(ns metabase.lib.util
+  "TODO -- this should probably be rolled into [[metabase.lib.query]]."
+  (:require
+   [clojure.set :as set]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.shared.util.i18n :as i18n]
+   [metabase.util.malli :as mu]))
+
+(defn- native-query->pipeline [query]
+  (merge {:lib/type :mbql/query
+          :type     :pipeline
+          ;; we're using `merge` here instead of threading stuff so the `:lib/` keys are the first part of the map for
+          ;; readability in the REPL.
+          :stages   [(merge (lib.options/ensure-uuid {:lib/type :mbql.stage/native})
+                            (set/rename-keys (:native query) {:query :native}))]}
+         (dissoc query :type :native)))
+
+(defn- inner-query->stages [{:keys [source-query], :as inner-query}]
+  (let [stages     (if source-query
+                     (inner-query->stages source-query)
+                     [])
+        stage-type (if (:native inner-query)
+                     :mbql.stage/native
+                     :mbql.stage/mbql)
+        ;; we're using `merge` here instead of threading stuff so the `:lib/` keys are the first part of the map for
+        ;; readability in the REPL.
+        this-stage (merge (lib.options/ensure-uuid
+                           {:lib/type stage-type})
+                          (dissoc inner-query :source-query))]
+    (conj stages this-stage)))
+
+(defn- mbql-query->pipeline [query]
+  (merge {:lib/type :mbql/query
+          :type     :pipeline
+          :stages   (inner-query->stages (:query query))}
+         (dissoc query :type :query)))
+
+(mu/defn pipeline :- ::lib.schema/query
+  "Take a 'traditional' MBQL query and convert it to a `:pipeline` query."
+  [query :- [:map [:type [:keyword]]]]
+  (condp = (:type query)
+    :pipeline query
+    :native   (native-query->pipeline query)
+    :query    (mbql-query->pipeline query)))
+
+(mu/defn ^:private non-negative-stage-index :- ::lib.schema.common/int-greater-than-or-equal-to-zero
+  "If `stage-number` index is a negative number e.g. `-1` convert it to a positive index so we can use `nth` on
+  `stages`. `-1` = the last stage, `-2` = the penultimate stage, etc."
+  [stages       :- [:sequential ::lib.schema/stage]
+   stage-number :- :int]
+  (let [stage-number' (if (neg? stage-number)
+                        (+ (count stages) stage-number)
+                        stage-number)]
+    (when (or (> stage-number' (dec (count stages)))
+              (neg? stage-number'))
+      (throw (ex-info (i18n/tru "Stage {0} does not exist" stage-number)
+                      {})))
+    stage-number'))
+
+(defn has-stage?
+  "Whether the query has a stage with `stage-number` (can be zero-indexed)."
+  [{:keys [stages], :as _query} stage-number]
+  ;; TODO -- this is a little bit duplicated from the logic above... find a way to consolidate?
```

Comment:
Something like `valid-stage` which would work similarly to `not-empty`, i.e. return the non-negative stage index if it's valid, `nil` otherwise?

Reply:
Yup, I already fixed this in #28717